### PR TITLE
Memory optimisation & rubocop on tests

### PR DESCRIPTION
**Pull request by abrisse**
The PR improves the memory management. On my benchmark (JSON-LD framing) the memory allocation is reduced by **37%**. 

It contains minor code modifications on tests to fix some rubocop offenses.

## Before

allocated memory by gem
-----------------------------------
  83.70 MB  json-ld

allocated memory by file
-----------------------------------
  38.59 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/frame.rb
  17.20 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/context.rb
  15.30 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/flatten.rb
   4.22 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/from_rdf.rb
   3.48 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/utils.rb

## After

allocated memory by gem
-----------------------------------
  52.95 MB  json-ld

allocated memory by file
-----------------------------------
  20.53 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/frame.rb
  13.39 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/context.rb
   6.42 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/flatten.rb
   4.22 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/from_rdf.rb
   3.48 MB  /home/aymeric/dev/gems/json-ld/lib/json/ld/utils.rb

### CODE REVIEW

File: lib/json/ld/context.rb [hunk 1]
```diff
@@ -1544,10 +1544,10 @@ def compact_iri(iri, base: nil, reverse: false, value: nil, vocab: nil)
         candidates = []
 
         term_definitions.each do |term, td|
-          next if td.nil? || td.id.nil? || td.id == iri || !iri.start_with?(td.id)
-
           # Skip term if `@prefix` is not true in term definition
-          next unless td.prefix?
+          next unless td&.prefix?
```
Comment (gkellogg) [inline lib/json/ld/context.rb:1548]: Learn something new every day 😄

**Comment by gkellogg** [comment]
There should probably either be a separate rubocop GH Action, or it should be added to .github/workflows/CI.yml.

File: spec/reader_spec.rb [hunk 2]
```diff
@@ -1,34 +1,20 @@
...
-  after(:each) {|example| puts logger.to_s if example.exception}
+  after { |example| puts logger if example.exception }
 
   it_behaves_like 'an RDF::Reader' do
-    let(:reader_input) {File.read(doap)}
-    let(:reader) {JSON::LD::Reader.new(reader_input)}
-    let(:reader_count) {doap_count}
-  end
-
-  describe ".for" do
```
Comment (gkellogg) [inline spec/reader_spec.rb:19]: Why remove these tests? are they redundant?

**Review by gkellogg — APPROVED** [review]
Changes look good, a few comments spread around to consider either for this PR or another.

File: lib/json/ld/context.rb (hunk 1 above)
Comment (abrisse) [reply lib/json/ld/context.rb:1548]: Quite practical, isn't it?

File: spec/reader_spec.rb (hunk 2 above)
Comment (abrisse) [reply spec/reader_spec.rb:19]: It's a mistake, thanks for pointing it out. I have rebased the branch.

**Pull request merged**